File: frontend/src/api/campaign.ts
import apiClient from './index'

// ─── Types ────────────────────────────────────────────────────────────────────

/** 活动类型：default=默认活动，community_care=社区成员关怀，developer_care=开发者关怀；旧版保留兼容 */
export type CampaignType =
  | 'default'
  | 'community_care'
  | 'developer_care'
  | 'promotion'
  | 'care'
  | 'invitation'
  | 'survey'

export interface CampaignListItem {
  id: number
  community_id: number | null
  name: string
  type: CampaignType
  status: string
  start_date: string | null
  end_date: string | null
  created_at: string
}

export interface CampaignDetail extends CampaignListItem {
  description: string | null
  owner_ids: number[]
  updated_at: string
}

export interface CampaignCreate {
  name: string
  type: CampaignType
  community_id?: number | null
  description?: string | null
  owner_ids?: number[]
  start_date?: string | null
  end_date?: string | null
}

export interface CampaignUpdate extends Partial<CampaignCreate> {
  status?: string
}

export interface PersonSnapshot {
  id: number
  display_name: string
  company: string | null
  email: string | null
  github_handle: string | null
}

export interface ContactOut {
  id: number
  campaign_id: number
  person_id: number
  status: string
  channel: string | null
  added_by: string
  last_contacted_at: string | null
  notes: string | null
  assigned_to_id: number | null
  person: PersonSnapshot | null
}

export interface PaginatedContacts {
  items: ContactOut[]
  total: number
  page: number
  page_size: number
}

export interface ActivityOut {
  id: number
  campaign_id: number
  person_id: number
  action: string
  content: string | null
  outcome: string | null
  operator_id: number | null
  created_at: string
}

export interface CampaignFunnel {
  pending: number
  contacted: number
  blocked: number
  total: number
}

export interface BulkStatusUpdate {
  contact_ids: number[]
  status: string
  notes?: string | null
}

export interface CommitteeSimple {
  id: number
  name: string
  member_count: number
}

export interface CsvImportResult {
  created: number
  matched: number
  skipped: number
  errors: string[]
}

// ─── Campaign CRUD ────────────────────────────────────────────────────────────

export async function listCampaigns(params?: { type?: string; status?: string }) {
  const res = await apiClient.get<CampaignListItem[]>('/campaigns', { params })
  return res.data
}

export async function createCampaign(data: CampaignCreate) {
  const res = await apiClient.post<CampaignDetail>('/campaigns', data)
  return res.data
}

export async function getCampaign(id: number) {
  const res = await apiClient.get<CampaignDetail>(`/campaigns/${id}`)
  return res.data
}

export async function updateCampaign(id: number, data: CampaignUpdate) {
  const res = await apiClient.patch<CampaignDetail>(`/campaigns/${id}`, data)
  return res.data
}

export async function deleteCampaign(id: number) {
  await apiClient.delete(`/campaigns/${id}`)
}

export async function getCampaignFunnel(id: number) {
  const res = await apiClient.get<CampaignFunnel>(`/campaigns/${id}/funnel`)
  return res.data
}

// ─── Contacts ─────────────────────────────────────────────────────────────────

export async function listContacts(
  campaignId: number,
  params?: { status?: string; page?: number; page_size?: number },
) {
  const res = await apiClient.get<PaginatedContacts>(
    `/campaigns/${campaignId}/contacts`,
    { params },
  )
  return res.data
}

export async function addContact(
  campaignId: number,
  data: { person_id: number; channel?: string; notes?: string },
) {
  const res = await apiClient.post<ContactOut>(`/campaigns/${campaignId}/contacts`, data)
  return res.data
}

export async function updateContactStatus(
  campaignId: number,
  contactId: number,
  data: { status: string; channel?: string; notes?: string },
) {
  const res = await apiClient.patch<ContactOut>(
    `/campaigns/${campaignId}/contacts/${contactId}/status`,
    data,
  )
  return res.data
}

export async function bulkUpdateContactStatus(
  campaignId: number,
  data: BulkStatusUpdate,
) {
  const res = await apiClient.patch<{ updated: number }>(
    `/campaigns/${campaignId}/contacts/bulk-status`,
    data,
  )
  return res.data
}

export async function deleteContact(campaignId: number, contactId: number) {
  await apiClient.delete(`/campaigns/${campaignId}/contacts/${contactId}`)
}

export async function importFromEvent(
  campaignId: number,
  data: { event_id: number; channel?: string },
) {
  const res = await apiClient.post<{ created: number; skipped: number }>(
    `/campaigns/${campaignId}/contacts/import-event`,
    data,
  )
  return res.data
}

export async function importFromPeople(
  campaignId: number,
  data: { person_ids: number[]; channel?: string },
) {
  const res = await apiClient.post<{ created: number; skipped: number }>(
    `/campaigns/${campaignId}/contacts/import-people`,
    data,
  )
  return res.data
}

// ─── Committee Import（社区成员关怀专用）──────────────────────────────────────

export async function listAvailableCommittees(campaignId: number) {
  const res = await apiClient.get<CommitteeSimple[]>(
    `/campaigns/${campaignId}/available-committees`,
  )
  return res.data
}

export async function importFromCommittees(
  campaignId: number,
  data: { committee_ids: number[]; channel?: string; assigned_to_id?: number | null },
) {
  const res = await apiClient.post<{ created: number; skipped: number }>(
    `/campaigns/${campaignId}/contacts/import-committee`,
    data,
  )
  return res.data
}

// ─── CSV Import（developer_care / community_care 共用）───────────────────────

export async function importFromCsv(campaignId: number, file: File) {
  const form = new FormData()
  form.append('file', file)
  const res = await apiClient.post<CsvImportResult>(
    `/campaigns/${campaignId}/contacts/import-csv`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  )
  return res.data
}

// ─── Activities ───────────────────────────────────────────────────────────────

export async function listActivities(campaignId: number, contactId: number) {
  const res = await apiClient.get<ActivityOut[]>(
    `/campaigns/${campaignId}/contacts/${contactId}/activities`,
  )
  return res.data
}

export async function addActivity(
  campaignId: number,
  contactId: number,
  data: { action: string; content?: string; outcome?: string },
) {
  const res = await apiClient.post<ActivityOut>(
    `/campaigns/${campaignId}/contacts/${contactId}/activities`,
    data,
  )
  return res.data
}

// ─── Campaign Task ─────────────────────────────────────────────────────────────

export type TaskStatus = 'not_started' | 'in_progress' | 'completed' | 'blocked'
export type TaskPriority = 'low' | 'medium' | 'high'

export interface CampaignTaskOut {
  id: number
  campaign_id: number
  title: string
  description: string | null
  status: TaskStatus
  priority: TaskPriority
  assignee_ids: number[]
  deadline: string | null
  created_by_id: number | null
  created_at: string
  updated_at: string
}

export interface CampaignTaskCreate {
  title: string
  description?: string | null
  status?: TaskStatus
  priority?: TaskPriority
  assignee_ids?: number[]
  deadline?: string | null
}

export interface CampaignTaskUpdate {
  title?: string
  description?: string | null
  status?: TaskStatus
  priority?: TaskPriority
  assignee_ids?: number[]
  deadline?: string | null
}

export async function listCampaignTasks(campaignId: number) {
  const res = await apiClient.get<CampaignTaskOut[]>(`/campaigns/${campaignId}/tasks`)
  return res.data
}

export async function createCampaignTask(campaignId: number, data: CampaignTaskCreate) {
  const res = await apiClient.post<CampaignTaskOut>(`/campaigns/${campaignId}/tasks`, data)
  return res.data
}

export async function updateCampaignTask(
  campaignId: number,
  taskId: number,
  data: CampaignTaskUpdate,
) {
  const res = await apiClient.patch<CampaignTaskOut>(
    `/campaigns/${campaignId}/tasks/${taskId}`,
    data,
  )
  return res.data
}

export async function deleteCampaignTask(campaignId: number, taskId: number) {
  await apiClient.delete(`/campaigns/${campaignId}/tasks/${taskId}`)
}

File: backend/app/schemas/campaign.py
from datetime import date, datetime

from pydantic import BaseModel

# ─── Campaign ─────────────────────────────────────────────────────────────────
# 活动类型说明：
#   default          — 默认活动（日常工作量/时间记录，节日海报推送等）
#   community_care   — 社区成员关怀（委员会委员等，支持从委员会导入、Excel导入）
#   developer_care   — 开发者关怀（面向大量人群，支持Excel批量导入）
#   promotion        — 推广宣传（旧版，保留兼容）
#   care             — 关怀回访（旧版，保留兼容）
#   invitation       — 邀请加入（旧版，保留兼容）
#   survey           — 问卷调研（旧版，保留兼容）

class CampaignCreate(BaseModel):
    name: str
    type: str  # default / community_care / developer_care / promotion / care / invitation / survey
    community_id: int | None = None
    description: str | None = None
    owner_id: int | None = None
    start_date: date | None = None
    end_date: date | None = None


class CampaignUpdate(BaseModel):
    name: str | None = None
    type: str | None = None
    description: str | None = None
    status: str | None = None   # draft / active / completed / archived
    owner_id: int | None = None
    start_date: date | None = None
    end_date: date | None = None


class CampaignListOut(BaseModel):
    id: int
    community_id: int | None
    name: str
    type: str
    status: str
    start_date: date | None
    end_date: date | None
    created_at: datetime

    model_config = {"from_attributes": True}


class CampaignOut(CampaignListOut):
    description: str | None
    owner_id: int | None
    updated_at: datetime

    model_config = {"from_attributes": True}


# ─── Campaign Contact ──────────────────────────────────────────────────────────

class ContactCreate(BaseModel):
    person_id: int
    channel: str | None = None
    notes: str | None = None
    assigned_to_id: int | None = None


class ContactStatusUpdate(BaseModel):
    status: str          # pending/contacted/responded/converted/declined
    channel: str | None = None
    notes: str | None = None
    assigned_to_id: int | None = None


class PersonSnapshot(BaseModel):
    id: int
    display_name: str
    company: str | None
    email: str | None
    github_handle: str | None

    model_config = {"from_attributes": True}


class ContactOut(BaseModel):
    id: int
    campaign_id: int
    person_id: int
    status: str
    channel: str | None
    added_by: str
    last_contacted_at: datetime | None
    notes: str | None
    assigned_to_id: int | None
    person: PersonSnapshot | None = None

    model_config = {"from_attributes": True}


class PaginatedContacts(BaseModel):
    items: list[ContactOut]
    total: int
    page: int
    page_size: int


# ─── Campaign Activity ─────────────────────────────────────────────────────────

class ActivityCreate(BaseModel):
    action: str   # sent_email/made_call/sent_wechat/in_person_meeting/got_reply/note
    content: str | None = None
    outcome: str | None = None


class ActivityOut(BaseModel):
    id: int
    campaign_id: int
    person_id: int
    action: str
    content: str | None
    outcome: str | None
    operator_id: int | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Funnel / Stats ───────────────────────────────────────────────────────────

class CampaignFunnel(BaseModel):
    pending: int = 0
    contacted: int = 0
    total: int = 0


# ─── Bulk Import ──────────────────────────────────────────────────────────────

class BulkImportFromEvent(BaseModel):
    event_id: int
    channel: str | None = None
    assigned_to_id: int | None = None


class BulkImportFromPeople(BaseModel):
    person_ids: list[int]
    channel: str | None = None
    assigned_to_id: int | None = None


# ─── Committee Import ─────────────────────────────────────────────────────────

class CommitteeSimple(BaseModel):
    """用于列举可导入的委员会"""
    id: int
    name: str
    member_count: int = 0

    model_config = {"from_attributes": True}


class BulkImportFromCommittees(BaseModel):
    """从一个或多个委员会批量导入成员，自动去重"""
    committee_ids: list[int]
    channel: str | None = None
    assigned_to_id: int | None = None


# ─── CSV/Excel Import ─────────────────────────────────────────────────────────
# 前端上传 CSV 文件，每行代表一个人，必填列：display_name
# 可选列：email, phone, company, github_handle, notes

class CsvImportResult(BaseModel):
    created: int = 0
    matched: int = 0
    skipped: int = 0
    errors: list[str] = []


# ─── Campaign Task ─────────────────────────────────────────────────────────────────

VALID_TASK_STATUSES = {"not_started", "in_progress", "completed", "blocked"}
VALID_TASK_PRIORITIES = {"low", "medium", "high"}


class CampaignTaskCreate(BaseModel):
    title: str
    description: str | None = None
    status: str = "not_started"  # not_started / in_progress / completed / blocked
    priority: str = "medium"    # low / medium / high
    assignee_ids: list[int] = []
    deadline: date | None = None


class CampaignTaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: str | None = None
    priority: str | None = None
    assignee_ids: list[int] | None = None
    deadline: date | None = None


class CampaignTaskOut(BaseModel):
    id: int
    campaign_id: int
    title: str
    description: str | None
    status: str
    priority: str
    assignee_ids: list[int]
    deadline: date | None
    created_by_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
